### ISSUE TITLE
some questions about function p_y_xz in node_model.py

### ISSUE THREAD

**Issue by leoncesc** [StanfordASL/Trajectron-plus-plus#1]
Thanks for your sharing great works!
I have some questions when I read your code.

1. In function p_y_xz in node_model.py file,  the input to the decoder has different shape with different mode (ModeKeys.TRAIN, ModeKeys.EVAL, ModeKeys.PREDICT). 
Under ModeKeys.TRAIN mode, initial input is:
`                input_ = torch.cat([zx, tensor_dict['joint_present'].repeat(k, 1)], dim=1)`[here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L821)
normal input is:
 `                   input_ = torch.cat([zx, dec_inputs], [dim=1)` [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L846)
 Under ModeKeys.EVAL mode, the initial input is:
`                    inputs = torch.cat([zx_time_tiled, dec_inputs.repeat(k, 1, 1)],
                                       dim=2)`[here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L871)
normal input is :
`                    inputs = torch.cat([zx_time_tiled, dec_inputs.repeat(k, 1, 1)],
                                       dim=2)` [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L871)

      Under ModeKeys.PREDICT mode, the initial input is:
`            input_ = torch.cat([zx, tensor_dict["joint_present"].repeat(k, 1)], dim=1)`[here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L888)
normal input is:
`                input_ = torch.cat([zx, dec_inputs],dim=1)` [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L938)
My question is what is the difference between different input under different mode, and why? 
What is in `tensor_dict['joint_present']` ?

2.  Under ModeKeys.TRAIN mode:
`                    mask = td.Bernoulli(probs=self.dec_sample_model_prob).sample((y_t.size()[0], 1))
                    y_t = mask * y_t + (1 - mask) * (tensor_dict[our_future][:, j, :].repeat(k, 1))` [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L833)
Under ModeKeys.PREDICT mode:
`                        ll_score = gmm_i.log_prob(search_grid).squeeze()
                        y_t_list.append(search_grid[0, torch.argmax(ll_score, dim=0)])` [here
](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L920)
what is the reason for doing this on `y_t` under different mode?
Where `self.dec_sample_model_prob` is assigned?

3.  what is the meaning of `log_pi_t, mu_t, log_sigma_t, corr_t`? Is each element in these variables corresponding to on GMM component?

**Comment by BorisIvanovic** [comment]
Hi! Thanks for reading through the code, and specifically the p_y_xz function. It's one of our workhorse methods and is unfortunately pretty complicated...

1. `tensor_dict['joint_present']` is defined [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L423), but specifically it's the node's current velocity (apologies for the poor naming).

As for the different inputs under different modes, here's what's going on
ModeKeys.TRAIN: The initial input is the encoded state and latent vector, concatenated with the agent's current velocity. The normal input after that gets its dec_inputs from [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L844), and is exactly the same velocity as before, but with a twist. The twist is that [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L834) there's a chance that we don't sample the GRU's output (I'm sorry that it says LSTM in the variable name, you'll have to excuse our variable naming...) but instead sample the ground truth output. This helps the model along during early training iterations when the decoder is sill mostly random.

ModeKeys.EVAL: It's the exact same idea here, although we always pass in the ground truth in order to get an estimate of our model's 1-step prediction accuracy. The reason we have to repeat the dec_inputs is because we sample from our model during eval, so we need to repeat by the number of samples we are using. Note that for any paper results, we actually don't use ModeKeys.EVAL, rather instead using ModeKeys.PREDICT (as you can see in our [evaluation script](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/notebooks/model_to_metric_nuScenes.py)). Even if we do call a model's eval_loss function, you can see that we pass in the ModeKeys.PREDICT value into p_y_xz [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L1142).

ModeKeys.PREDICT: Very similar to ModeKeys.TRAIN for the initial value. However, the regular value (`dec_inputs = y_t`) is just what we get after running the previous GRU cell (you can see this [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L925)).

2. `self.dec_sample_model_prob` is assigned [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L287). 

The reason we have different treatments for `y_t` is because for ModeKeys.TRAIN, we want to either pass in the previous GRU cell's output or the ground truth (hence the masking by a Bernoulli random variable). For ModeKeys.PREDICT (and the specific line you're pointing to is the GMM mode output of our model, starting at [this](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L900) if statement), we need to find the most likely part of the GMM and return that as our model's output (hence the argmax of the log-likelihood). For plain predictions (i.e. not the GMM mode), you can see we just sample from the GMM [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/node_model.py#L925).

3. `log_pi_t, mu_t, log_sigma_t, corr_t` are each different parts of a GMM. Specifically, `log_pi_t` is the logarithm of the scalar weighting value for the each GMM component at time t (this means one of the tensor dimensions is the number of GMM components). `mu_t` are the means at each timestep t. `log_sigma_t, corr_t` are values that we use to form the GMM's covariance matrix at each timestep t. For more details, you can see how these values are used to form a GMM [here](https://github.com/StanfordASL/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/components/gmm2d.py#L7).

I hope this helps!

**Comment by BorisIvanovic** [comment]
Closing, please feel free to reopen if you have any further questions!

**Comment by leoncesc** [comment]
Thanks for your help! @BorisIvanovic 
Another question is the calculation of reconstruction loss(log likelihood) in function `log_prob` in gmm2d.py [here](https://github.com/leoncesc/Trajectron-plus-plus/blob/3588a3ebea472623dce537a528801ac6b441fac2/code/model/components/gmm2d.py#L49), 
```
z = (torch.sum((dx/self.sigmas)**2, dim=-1) -
             2*self.corrs*torch.prod(dx, dim=-1)/torch.prod(self.sigmas, dim=-1))    # [..., GMM_c]
```
       
```
 component_log_p = -(torch.log(self.one_minus_rho2) + 2*torch.sum(self.log_sigmas, dim=-1) 
                           + z/self.one_minus_rho2 +
                            2*np.log(2*np.pi))/2
```

According to the Bivariate case [Wikepedia](https://en.wikipedia.org/wiki/Multivariate_normal_distribution) I think it is minus `z` as shown below
```
component_log_p = -(torch.log(self.one_minus_rho2) + 2*torch.sum(self.log_sigmas, dim=-1) 
                           - z/self.one_minus_rho2 +
                            2*np.log(2*np.pi))/2
```
